Make persists concurrent with adding rows in batch ingestion

loquisgon opened the pull request:
Recent testing have shown a potential performance degradation in the bounded memory work for segment creation of batch ingestion.  The reason is most probably the fact that intermediate persists are blocking, serial, with ingestion. We will add back concurrent persists and test performance to validate that previous performance is back.

Changes:
Intermediate persists are made concurrent with adding rows in the batch appenderator. 

We are also interested in supporting three main modes for the batch appenderator, this PR is implementing also the following requirements:

-  Fix the performance regression by making persists & push concurrent 
- Have a rollback flag called batchProcessingMode. This configuration setting will be able to take three string values: LEGACY, CLOSED_SEGMENTS, CLOSED_SEGMENTS_SINKS (see table below). The default value will be: CLOSED_SEGMENTS. Note that most existing tests will exercise the default but there are specific tests for the other two modes (see configuration docs to know what each of these mean).
- Update docs to document the new processing mode

This PR has:
- [ X] been self-reviewed.
   - [x] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [X ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [X ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

loquisgon commented:
I replaced the semaphore using the style of concurrency used in the appenderator. Now persists and push are ran in the background.

jihoonson commented on server/src/test/java/org/apache/druid/segment/realtime/appenderator/BatchAppenderatorTest.java:
```diff
@@ -806,7 +831,8 @@ public void testTotalRowCount() throws Exception
 
-      appenderator.persistAll(null).get();
+      appenderator.persistAll(null);
+      waitForPersists();
```
Why not `persistAll(null).get()`?

loquisgon replied:
Typo --- fixed in next commit.

jihoonson commented on server/src/test/java/org/apache/druid/segment/realtime/appenderator/BatchAppenderatorTest.java:
```diff
@@ -913,6 +939,10 @@ static InputRow createInputRow(String ts, String dim, Object met)
 
+  private void waitForPersists() throws InterruptedException
+  {
+    Thread.sleep(500);
```
Sleeps are bad. These will make the unit testing slower. Also, I bet all the sleeps you added will make these tests quite flaky.

loquisgon replied:
The use case here is somehow wait for a concurrent task to check on its effect, it seems like a legitimate use for sleep. Let me know what you are thinking that could be better.

loquisgon replied:
oh I see your suggestion below...

jihoonson commented on server/src/test/java/org/apache/druid/segment/realtime/appenderator/BatchAppenderatorTest.java:
```diff
@@ -382,12 +387,14 @@ public void testTaskDoesNotFailAsExceededMemoryWithSkipBytesInMemoryOverheadChec
 
       appenderator.startJob();
       appenderator.add(IDENTIFIERS.get(0), createInputRow("2000", "foo", 1), null);
+      waitForPersists();
```
I think you could keep the future of persist triggered in `add()` in a variable. Then you can add a method used only for testing that returns the persist future. Then you can finally wait for the future to be done instead of sleeping.

loquisgon replied:
I was trying to avoid more changes to the class (exposing methods etc) but I can pursue your suggestion

jihoonson commented:
BTW, I assumed that the core logic is not changed and reviewed only the concurrency part. Let me know if this is not true.

loquisgon commented on server/src/main/java/org/apache/druid/segment/realtime/appenderator/BatchAppenderator.java:
```diff
@@ -158,6 +169,11 @@
     maxBytesTuningConfig = tuningConfig.getMaxBytesInMemoryOrDefault();
     skipBytesInMemoryOverheadCheck = tuningConfig.isSkipBytesInMemoryOverheadCheck();
+    if (tuningConfig.getMaxPendingPersists() < 1) {
+      maxPendingPersists = DEFAULT_PENDING_PERSISTS;
```
It just seems reasonable to me... I can go back to the previous default, I have no issues with that.

loquisgon commented:
You are correct, @jihoonson, core logic is not changed.

loquisgon commented on server/src/main/java/org/apache/druid/segment/realtime/appenderator/BatchAppenderator.java:
```diff
@@ -177,9 +193,53 @@ public Object startJob()
+
+  private void initializeExecutors()
+  {
+    log.info("There will be up to[%d] pending persists", maxPendingPersists);
```
It is debug now

loquisgon commented on server/src/main/java/org/apache/druid/segment/realtime/appenderator/BatchAppenderator.java:
```diff
@@ -110,20 +113,28 @@
 
   // This variable updated in add(), persist(), and drop()
-  private int rowsCurrentlyInMemory = 0;
+  private final AtomicInteger rowsCurrentlyInMemory = new AtomicInteger();
```
Removed unnecessary atomics in next committ.

loquisgon commented on server/src/main/java/org/apache/druid/segment/realtime/appenderator/BatchAppenderator.java:
```diff
@@ -110,20 +113,28 @@
+  private volatile Throwable persistError;
+
+
+  private final ConcurrentHashMap<SegmentIdWithShardSpec, Sink> sinks = new ConcurrentHashMap<>();
```
No concurrency control required for this map, removed in next commit.

loquisgon commented:
Changes I added in the latest commit: 

- Make sure close() waits for persists & pushes before closing
- Merged master branch to pick up fixes to a previous issue (serialization/deserialization of JSON interval when it contains UTC timestamps)
- Renamed `BatchAppenderator#clearSinkMetadata` to `BatchAppenderator#clearSinkMemoryCountersAndDiskStoredData` to make intent clearer

loquisgon commented on server/src/main/java/org/apache/druid/segment/realtime/appenderator/UnifiedIndexerAppenderatorsManager.java:
```diff
@@ -227,8 +226,81 @@ public Appenderator createOfflineAppenderatorForTask(
+          DatasourceBundle::new
+      );
+
+      Appenderator appenderator = Appenderators.createLegacyOffline(
```
This is wrong, it should be createClosedSegmentsOffline, I will fix it

loquisgon replied:
Done

jon-wei commented on indexing-service/src/main/java/org/apache/druid/indexing/common/config/TaskConfig.java:
```diff
@@ -43,6 +44,15 @@
       "org.apache.hadoop:hadoop-client:2.8.5"
   );
 
+  public enum BatchProcesingMode
```
BatchProcesingMode -> BatchProcessingMode

imply-jhan commented on docs/configuration/index.md:
```diff
@@ -1343,7 +1343,8 @@ Additional peon configs include:
 |`druid.indexer.task.baseDir`|Base temporary working directory.|`System.getProperty("java.io.tmpdir")`|
 |`druid.indexer.task.baseTaskDir`|Base temporary working directory for tasks.|`${druid.indexer.task.baseDir}/persistent/task`|
-|`druid.indexer.task.useLegacyBatchProcessing`|If false, native batch ingestion will use a new, recommended, code path with memory optimized code for the segment creation phase. If true it will use the previous code path for the create segments phase of batch ingestion. This does not apply to streaming ingestion, just to batch. This setting should only be used when a bug is suspected or found in  ...
+|`druid.indexer.task.batchMemoryMappedIndex`|DEPRECATED: Use `druid.indexer.task.batchProcessingMode` instead. If false, native batch ingestion will use a new, recommended, code path with memory optimized code for the segment creation phase. If true it will use the previous code path for the create segments phase of batch ingestion. This does not apply to streaming ingestion, just to batch. This  ...
```
Maybe call out that set batchMemoryMappedIndex to true will set batchProcessingMode to LEGACY and overwrite the batchProcessingMode value.

imply-jhan commented on indexing-service/src/main/java/org/apache/druid/indexing/common/config/TaskConfig.java:
```diff
@@ -117,7 +131,19 @@ public TaskConfig(
+    } else {
+      // batchProcessingMode input string is invalid, just use the default...log message somewhere???
+      this.batchProcessingMode = BatchProcesingMode.CLOSED_SEGMENTS; // Default
+    }
```
Can we log the batchProcessingMode value after this if block?

loquisgon replied:
yeah

clintropolis commented on indexing-service/src/test/java/org/apache/druid/indexing/common/task/IngestionTestBase.java:
```diff
@@ -314,7 +314,8 @@ public File getTaskReportsFile()
         final TaskToolbox box = new TaskToolbox(
-            new TaskConfig(null, null, null, null, null, false, null, null, null, false, false),
+            new TaskConfig(null, null, null, null, null, false, null, null, null, false, false,
+                           TaskConfig.BatchProcessingMode.CLOSED_SEGMENTS.name()),
```
nit: should this use default instead of this explicit value?

loquisgon replied:
done in next

clintropolis commented on server/src/main/java/org/apache/druid/segment/realtime/appenderator/AppenderatorImpl.java:
```diff
@@ -0,0 +1,1596 @@
+
+/**
+ * This class is to support LEGACY and CLOSED_SEGMENTS appenderators. It is copied as-is
+ * from 0.21 and it is meant to keep for backward compatibility. For now though this class
```
nit: this isn't actually a copy of 0.21... its of some intermediary state between 0.21 and 0.22. I think `StreamAppenderator` is technically a copy of `ApenderatorImpl` in 0.21.

loquisgon replied:
Updated in next

clintropolis commented on server/src/test/java/org/apache/druid/segment/realtime/appenderator/LegacyAndClosedSegmentsAppenderatorTester.java:
```diff
@@ -0,0 +1,293 @@
+import java.util.Map;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+public class LegacyAndClosedSegmentsAppenderatorTester implements AutoCloseable
```
super-nit: `OpenAndClosed...` instead of `LegacyAndClosed`

loquisgon replied:
yeah

clintropolis commented on server/src/test/java/org/apache/druid/segment/realtime/appenderator/LegacyAndClosedSegmentsBatchAppenderatorDriverTest.java:
```diff
@@ -0,0 +1,205 @@
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class LegacyAndClosedSegmentsBatchAppenderatorDriverTest extends EasyMockSupport
```
same super nit replacing Legacy with Open

loquisgon replied:
check

clintropolis commented on server/src/test/java/org/apache/druid/segment/realtime/appenderator/LegacyAndClosedSegmentsBatchAppenderatorTest.java:
```diff
@@ -0,0 +1,229 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class LegacyAndClosedSegmentsBatchAppenderatorTest extends InitializedNullHandlingTest
```
same nit legacy/open

loquisgon replied:
check

clintropolis commented on docs/configuration/index.md:
```diff
@@ -1343,7 +1343,7 @@ Additional peon configs include:
 |`druid.indexer.task.baseDir`|Base temporary working directory.|`System.getProperty("java.io.tmpdir")`|
 |`druid.indexer.task.baseTaskDir`|Base temporary working directory for tasks.|`${druid.indexer.task.baseDir}/persistent/task`|
-|`druid.indexer.task.useLegacyBatchProcessing`|If false, native batch ingestion will use a new, recommended, code path with memory optimized code for the segment creation phase. If true it will use the previous code path for the create segments phase of batch ingestion. This does not apply to streaming ingestion, just to batch. This setting should only be used when a bug is suspected or found in  ...
+|`druid.indexer.task.batchProcessingMode`| Batch ingestion tasks have three operating modes that control how intermediary segments are constructed and tracked: `OPEN_SEGMENTS`, `CLOSED_SEGMENTS`, and `CLOSED_SEGMENT_SINKS`. `OPEN_SEGMENTS` will use code based on the original batch ingestion path and performs a `mmap` on intermediary segments to build a timeline so that these segments can be query ...
```
```suggestion
|`druid.indexer.task.batchProcessingMode`| Batch ingestion tasks have three operating modes that control how intermediary segments are constructed and tracked: `OPEN_SEGMENTS`, `CLOSED_SEGMENTS`, and `CLOSED_SEGMENT_SINKS`. `OPEN_SEGMENTS` will use code based on the original batch ingestion path and performs a `mmap` on intermediary segments to build a timeline so that these segments can be queryable by realtime queries. This is not needed at all for batch, so the default mode, `CLOSED_SEGMENTS`, eliminates `mmap` of intermediary segments, but still tracks the entire set of segments in heap. The `CLOSED_SEGMENTS_SINKS` mode is the most aggressive and should have the smallest memory footprint, and works by eliminating in memory tracking and `mmap` of intermediary segments produced during segment creation. This mode isn't as well tested as other modes so is currently considered experimental. `OPEN_SEGMENTS` mode can be selected if any problems occur with the 2 newer modes. |`CLOSED_SEGMENTS`|
```
should fix spellcheck CI failure i think
